owltorrent-network/src/main/java/edu/rice/owltorrent/network/TorrentManager.java

```diff
@@ -0,0 +1,199 @@
+      } catch (IOException e) {
+        log.error(String.format("Error connecting peer id=%s", peer.getPeerID().toString()));
+      }
+    }
+  }
+
+  public void startDownloadingAsynchronously() {
+    // TODO: make thread pool scheduler
+    new Thread(this).start();
+  }
+
+  public float getProgressPercent() {
+    return completedPieces.size() * 1.0f / totalPieces;
+  }
+
+  // For handshake listener
+  // Later if get updated peerlist from tracker.
+  public void addPeer(PeerConnector connector, Peer peer) {
+    this.peers.put(peer, connector);
+  }
+
+  @Override
+  public void close() throws Exception {
+    for (var pair : peers.entrySet()) {
+      pair.getValue().close();
+    }
+  }
+
+  private void requestPieceFromPeer(Peer peer, int pieceNum, int blockNum, int blockSize) {
+    PeerConnector peerConnector = peers.get(peer);
+    if (peerConnector == null) {
+      return;
+    }
+
+    // TODO: fix int casting.
+    try {
+      peerConnector.writeMessage(
+          PieceActionMessage.makeRequestMessage(pieceNum, blockNum * blockSize, blockSize));
+    } catch (IOException e) {
+      log.error(e);
+    }
+  }
+
+  @Override
+  public void run() {
+    while (!uncompletedPieces.isEmpty()) {
+      // Request a missing piece from each Peer
+      List<Peer> connections = new ArrayList<>(peers.keySet());
+      Collections.shuffle(connections);
+
+      for (PieceStatus progress : uncompletedPieces.values()) {
+        for (int i = 0; i < progress.status.size(); i++) {
+          if (connections.isEmpty()) break;
+
+          if (progress.status.get(i).get() != BLOCK_NOT_STARTED) {
+            continue;
+          }
+
+          requestPieceFromPeer(connections.remove(0), progress.pieceIndex, i, progress.blockLength);
+        }
+        if (connections.isEmpty()) break;
+      }
+      // Thread.sleep(1000);
+    }
+  }
+
+  /**
+   * Reports the progress of a block
+   *
+   * @param blockInfo block info
+   * @return false if another thread already started to download that block or the block size is not
+   *     expected
+   */
+  public boolean reportAndValidatePieceInProgress(FileBlockInfo blockInfo) {
+    // TODO: fix int casting, and also add handling when piece length isn't a multiple of default
+    //  block num
+    PieceStatus status =
+        uncompletedPieces.computeIfAbsent(
+            blockInfo.getIndex(),
+            i ->
+                new PieceStatus(
+                    i, DEFAULT_BLOCK_NUM, (int) torrent.getPieceLength() / DEFAULT_BLOCK_NUM));
+    if (status.blockLength != blockInfo.getLength()) {
+      return false;
+    }
+
+    return status
+        .status
+        .get(blockInfo.getBegin() / status.blockLength)
+        .compareAndSet(BLOCK_NOT_STARTED, BLOCK_IN_PROGRESS);
+  }
+
+  public void reportPieceCompletion(FileBlockInfo blockInfo) {
+    PieceStatus status = uncompletedPieces.get(blockInfo.getIndex());
+    assert status != null;
+
+    if (status
```

Comment:
just do a set to BLOCK_DONE and move the for loop out of the if statement, it's simpler (and we really shouldn't every be setting a block to done we've already finished, if the concurrency is implemented correctly)

Reply:
I personally prefer to be defensive in these sorts of situations... Shouldn't be a increase in latency since most of the time there shouldn't be concurrent threads working on the same block

Reply:
I'm not worried about latency it's just a little confusing, but okay it's alright